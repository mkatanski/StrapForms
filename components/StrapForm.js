import React, { Component } from 'react'
import T from 'prop-types'

import { isValid, isArray } from './utils/helpers'

export const StrapFormContextTypes = {
  listenTo: T.func,
  dispatchEvent: T.func,
}

export const StrapFormPropTypes = {
  onInputBlur: T.func,
  onInputChange: T.func,
  onSubmit: T.func,
  onSubmitFail: T.func,
}

export default function (Form) {
  class StrapForm extends Component {
    static propTypes = {
      ...StrapFormPropTypes,
      children: T.any,
    }

    static defaultProps = {
      children: null,
      onInputChange: () => { },
      onInputBlur: () => { },
      onSubmit: () => { },
    }

    static childContextTypes = {
      ...StrapFormContextTypes,
    }

    constructor(...args) {
      super(...args)

      this.listeners = {}
      this.isSubmitting = false
      this.submitted = false
      this.isPristine = true
      this.values = {}
      this.errors = {}
      this.warnings = {}
      this.validating = []
    }

    getChildContext() {
      return {
        listenTo: this.listenTo,
        dispatchEvent: this.dispatchEvent,
      }
    }

    componentDidMount() {
      this.listenTo('onInputChange', this.handleOnInputChange)
      this.listenTo('onInputBlur', this.handleOnInputBlur)
      this.listenTo('onBeforeAsyncValidation', this.handleOnBeforeAsyncValidation)
      this.listenTo('onAfterAsyncValidation', this.handleOnAfterAsyncValidation)
    }

    get formData() {
      return {
        errors: this.errors,
        warnings: this.warnings,
        isValid: isValid(this.errors),
        isPristine: this.isPristine,
        isSubmitting: this.isSubmitting,
        isValidating: this.validating.length !== 0,
        values: this.values,
      }
    }

    getValueWithDefault = (value, defaultValue) => (value !== undefined ? value : defaultValue)

    getFormInputData = inputData => ({
      ...this.formData,
      inputName: inputData.inputName,
      value: inputData.value,
    })

    dispatchEvent = (eventName, eventData) => {
      const listenersResults = []

      if (!isArray(this.listeners[eventName])) {
        return undefined
      }

      this.listeners[eventName].forEach((listener) => {
        const result = listener(eventData)
        if (result !== undefined) {
          listenersResults.push(result)
        }
      })

      return listenersResults
    }

    listenTo = (eventName, func) => {
      if (!isArray(this.listeners[eventName])) {
        this.listeners[eventName] = []
      }

      this.listeners[eventName].push(func)
    }

    updateFormData = (formData) => {
      const fData = { ...formData }
      const inputsData = fData.inputsData

      this.isPristine = this.getValueWithDefault(fData.isPristine, this.isPristine)
      this.isSubmitting = this.getValueWithDefault(fData.isSubmitting, this.isSubmitting)
      this.submitted = this.getValueWithDefault(fData.submitted, this.submitted)

      if (isArray(inputsData) && inputsData.length !== 0) {
        inputsData.forEach((data) => {
          try {
            const { value, inputName, errors, warnings } = data
            this.errors[inputName] = errors
            this.warnings[inputName] = warnings
            this.values[inputName] = value
          } catch (e) {
            // handle error
          }
        })
      }

      this.dispatchEvent('onFormUpdate', this.formData)
    }

    handleOnBeforeAsyncValidation = (e) => {
      this.validating.push(e.inputName)
      this.dispatchEvent('onFormUpdate', this.formData)
    }

    handleOnAfterAsyncValidation = (e) => {
      const i = this.validating.indexOf(e.inputName)
      this.validating.splice(i, 1)
    }

    handleInput = (method, inputOptions, isPristine) => {
      this.updateFormData({
        isPristine: this.getValueWithDefault(isPristine, this.isPristine),
        inputsData: [inputOptions],
      })
      this.props[method](this.getFormInputData(inputOptions))
    }

    handleOnInputBlur = (inputOptions) => {
      this.handleInput('onInputBlur', inputOptions, false)
    }

    handleOnInputChange = (inputOptions) => {
      this.handleInput('onInputChange', inputOptions)
    }

    handleSubmit = async (event) => {
      event.preventDefault()
      this.updateFormData({ isSubmitting: true, submitted: true })

      // TODO: Add possibility to edit validators during this event
      const syncResult = this.dispatchEvent('onFormSyncValidate')
      this.updateFormData({ inputsData: syncResult })

      if (!isValid(this.errors)) {
        this.updateFormData({ isSubmitting: false })
        this.dispatchEvent('onFormSubmitFail', this.formData)
        if (this.props.onSubmitFail) {
          await this.props.onSubmitFail(this.formData)
        }
        return
      }

      const asyncHandlers = this.dispatchEvent('onFormAsyncValidate')

      let asyncResult = []
      if (asyncHandlers) {
        asyncResult = await Promise.all(asyncHandlers)
      }

      this.updateFormData({ inputsData: asyncResult })

      if (isValid(this.errors)) {
        this.dispatchEvent('onFormSubmit', this.formData)
        await this.props.onSubmit(this.formData)
      } else {
        this.dispatchEvent('onFormSubmitFail', this.formData)
        if (this.props.onSubmitFail) {
          await this.props.onSubmitFail(this.formData)
        }
      }

      this.updateFormData({ isSubmitting: false })
    }

    render() {
      return (
        <Form handleSubmit={this.handleSubmit}>
          {this.props.children}
        </Form>
      )
    }
  }

  return StrapForm
}
