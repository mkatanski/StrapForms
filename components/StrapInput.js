import React, { Component } from 'react'
import T from 'prop-types'

import { StrapFormContextTypes } from './StrapForm'
import { StrapGroupContextTypes } from './StrapGroup'

export const StrapInputPropTypes = {
  input: T.shape({
    disabled: T.bool,
    id: T.string.isRequired,
    name: T.string.isRequired,
    onBlur: T.func.isRequired,
    onChange: T.func.isRequired,
    readOnly: T.bool,
    value: T.any.isRequired,
  }),
  hasErrors: T.bool,
  hasWarnings: T.bool,
  errors: T.object,
  warnings: T.object,
  touched: T.bool,
  isValidating: T.bool,
}

export default function (Input) {
  class StrapInput extends Component {
    static contextTypes = {
      ...StrapFormContextTypes,
      ...StrapGroupContextTypes,
    }

    static propTypes = {
      name: T.string.isRequired,
      asyncValidation: T.func,
      children: T.any,
      disabled: T.bool,
      initialValue: T.any,
      readOnly: T.bool,
      validate: T.array,
      warn: T.array,
    }

    static defaultProps = {
      asyncValidation: null,
      children: null,
      disabled: false,
      readOnly: false,
      validate: [],
      warn: [],
      initialValue: '',
    }

    constructor(...args) {
      super(...args)

      this.asyncQueue = []
    }

    state = {
      touched: false,
      isValidating: false,
      errors: {},
      warnings: {},
      value: this.props.initialValue || '',
    }

    componentWillMount() {
      this.context.listenTo('onFormSyncValidate', this.handleOnFormSyncValidate)
      this.context.listenTo('onFormAsyncValidate', this.handleOnFormAsyncValidate)
      if (this.context.registerToGroup) {
        this.context.registerToGroup(this.props.name)
      }
    }

    componentWillUnmount() {
      if (this.context.deregisterFromGroup) {
        this.context.deregisterFromGroup(this.props.name)
      }
    }

    getValidationResult = validationResult => ({
      inputName: this.props.name,
      value: this.state.value,
      ...validationResult,
    })

    isValid = errors => Object.keys(errors).length === 0

    dispatchEvent = (eventName, data) => {
      const enhancedData = {
        inputName: this.props.name,
        ...data,
      }

      return this.context.dispatchEvent(eventName, enhancedData)
    }

    performAsyncValidation = async (value) => {
      const asyncValidateMethod = this.props.asyncValidation
      let message

      if (!asyncValidateMethod || this.props.disabled || this.props.readOnly) {
        return message
      }

      this.setState({ isValidating: true })
      this.asyncQueue.push(value)
      this.dispatchEvent('onBeforeAsyncValidation', {
        value,
        asyncQueue: [...this.asyncQueue],
      })

      try {
        await asyncValidateMethod(value, this.state.values)
      } catch (error) {
        message = error.message
      }

      const queueIndex = this.asyncQueue.indexOf(value)
      this.asyncQueue.splice(queueIndex, 1)

      this.dispatchEvent('onAfterAsyncValidation', {
        message,
        asyncQueue: [...this.asyncQueue],
      })

      this.setState({ isValidating: false })
      return message
    }

    performSyncValidation = (value) => {
      const errorValidators = [...this.props.validate]
      const warnValidators = [...this.props.warn]
      const { errors, warnings } = this.state

      if (this.props.disabled || this.props.readOnly) {
        return { errors, warnings }
      }

      this.dispatchEvent('onBeforeSyncValidation', {
        errors: [...errors],
        warnings: [...warnings],
        value,
        errorValidators,
        warnValidators,
      })

      this.validate({ validators: errorValidators, errors, value })
      this.validate({ validators: warnValidators, errors: warnings, value })

      this.dispatchEvent('onAfterSyncValidation', { errors, warnings, value })
      return { errors, warnings }
    }

    performFullValidation = async (value) => {
      let asyncError
      const validationResult = this.performSyncValidation(value)

      if (this.isValid(validationResult.errors)) {
        asyncError = await this.performAsyncValidation(value)
      }

      if (asyncError) {
        validationResult.errors.async = asyncError
      }

      return validationResult
    }

    validate = ({ validators, errors, value }) => {
      const inputErrors = errors || {}

      if (!validators) { return }

      validators.forEach((validator, index) => {
        try {
          const result = validator(value, this.state.values)
          if (result) {
            inputErrors[index] = result
          } else if (inputErrors[index]) {
            delete inputErrors[index]
          }
        } catch (e) {
          // catch error
        }
      })
    }

    handleOnChange = (e) => {
      const value = e.target ? e.target.value : e
      const errors = this.state.errors

      if (errors.async) {
        delete errors.async
      }

      this.setState({ value, errors })
      const validationResult = this.performSyncValidation(value)

      this.dispatchEvent('onInputChange', {
        value,
        isValid: this.isValid(validationResult.errors),
        ...validationResult,
      })

      this.setState({ ...validationResult })
    }

    handleOnBlur = async (e) => {
      const value = e.target ? e.target.value : e
      this.setState({ touched: true, value })

      const validationResult = await this.performFullValidation(value)
      const isValid = this.isValid(validationResult.errors)

      this.dispatchEvent('onInputBlur', { value, isValid, ...validationResult })
      this.setState({ ...validationResult })
    }

    handleOnFormAsyncValidate = async () => {
      const asyncError = await this.performAsyncValidation(this.state.value)
      const validationResult = { errors: {}, warnings: {} }

      if (asyncError) {
        validationResult.errors.async = asyncError
      }

      this.setState({ ...validationResult, touched: true })
      return this.getValidationResult(validationResult)
    }

    handleOnFormSyncValidate = () => {
      const validationResult = this.performSyncValidation(this.state.value)
      this.setState({ ...validationResult, touched: true })
      return this.getValidationResult(validationResult)
    }

    filterProps = () => {
      const externalProps = {}
      const propTypesKeys = Object.keys(this.constructor.propTypes)
      Object.keys(this.props).forEach((key) => {
        if (propTypesKeys.includes(key)) {
          return
        }
        externalProps[key] = this.props[key]
      })
      return externalProps
    }

    render() {
      const { errors, warnings, value } = this.state
      const props = {
        input: {
          disabled: this.props.disabled,
          id: this.props.name,
          name: this.props.name,
          onBlur: this.handleOnBlur,
          onChange: this.handleOnChange,
          readOnly: this.props.readOnly,
          value,
        },
        hasErrors: errors && !this.isValid(errors),
        hasWarnings: warnings && !this.isValid(warnings),
        errors,
        warnings,
        touched: this.state.touched,
        isValidating: this.asyncQueue.length > 0,
      }

      return (
        <Input {...this.filterProps()} {...props}>
          {this.props.children}
        </Input>
      )
    }
  }

  return StrapInput
}
