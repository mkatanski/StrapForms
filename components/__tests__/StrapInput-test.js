/* eslint-disable react/prefer-stateless-function,react/no-multi-comp */
import React, { Component } from 'react'
import T from 'prop-types'
import renderer from 'react-test-renderer'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import StrapInput from '../StrapInput'
import { StrapFormContextTypes } from '../StrapForm'

configure({ adapter: new Adapter() })

const fetchSimulate = ms => new Promise(resolve => setTimeout(resolve, ms))

class ContextWrapper extends Component {
  static propTypes = {
    children: T.any.isRequired,
  }

  static childContextTypes = {
    ...StrapFormContextTypes,
  }

  getChildContext() {
    return {
      listenTo: () => { },
      dispatchEvent: () => { },
    }
  }

  render() {
    return (
      <div>
        {this.props.children}
      </div>
    )
  }
}

class PureInput extends Component {
  render() {
    // eslint-disable-next-line
    const { input, className } = this.props

    return (
      <div>
        <input {...input} className={className} />
        <span>{JSON.stringify(this.props)}</span>
      </div>
    )
  }
}

const TestInput = StrapInput(PureInput)

describe('StrapInput', () => {
  const checkChange = async (done, method, options) => {
    const { val, validate, warn, expected, disabled, readOnly, asyncValidation } = options
    const mockListenTo = jest.fn()
    const mockDispatchEvent = jest.fn()

    const component = shallow(
      (
        <TestInput
          name="test_input_name"
          validate={validate}
          warn={warn}
          asyncValidation={asyncValidation}
          disabled={!!disabled}
          readOnly={!!readOnly}
        />
      ),
      {
        context: {
          listenTo: mockListenTo,
          dispatchEvent: mockDispatchEvent,
        },
      }
    )

    const instance = component.instance()

    expect(component.state('touched')).toBe(false)
    expect(component.state('value')).toBe('')
    expect(mockDispatchEvent.mock.calls.length).toBe(0)

    await instance[method](val)

    if (expected.asyncMessage) {
      expect(mockDispatchEvent.mock.calls.length).toBe(3)
      expect(mockDispatchEvent.mock.calls[0][0]).toBe('onBeforeAsyncValidation')
      expect(mockDispatchEvent.mock.calls[0][1]).toEqual({
        inputName: 'test_input_name',
        value: val,
      })
      expect(mockDispatchEvent.mock.calls[1][0]).toBe('onAfterAsyncValidation')
      expect(mockDispatchEvent.mock.calls[1][1]).toEqual({
        inputName: 'test_input_name',
        message: expected.asyncMessage,
      })

      expect(mockDispatchEvent.mock.calls[2][0]).toBe(expected.eventName)
      expect(mockDispatchEvent.mock.calls[2][1]).toEqual({
        errors: expected.errors,
        inputName: 'test_input_name',
        isValid: expected.isValid,
        value: val,
        warnings: expected.warnings,
      })
    } else {
      expect(mockDispatchEvent.mock.calls.length).toBe(1)
      expect(mockDispatchEvent.mock.calls[0][0]).toBe(expected.eventName)
      expect(mockDispatchEvent.mock.calls[0][1]).toEqual({
        errors: expected.errors,
        inputName: 'test_input_name',
        isValid: expected.isValid,
        value: val,
        warnings: expected.warnings,
      })
    }

    expect(component.state('touched')).toBe(expected.touched)
    expect(component.state('value')).toBe(val)
    done()
  }

  it('can be rendered without problems with default values', () => {
    const component = renderer.create(
      <ContextWrapper>
        <TestInput name="test_input_name" />
      </ContextWrapper>
    )

    const input = component.toJSON()
    expect(input).toMatchSnapshot()
  })

  it('can be rendered without problems with initial value', () => {
    const component = renderer.create(
      <ContextWrapper>
        <TestInput
          name="test_input_name"
          initialValue="init value"
          className="customProperty"
        />
      </ContextWrapper>
    )

    const input = component.toJSON()
    expect(input).toMatchSnapshot()
  })

  it('can be rendered without problems as disabled and readonly', () => {
    const component = renderer.create(
      <ContextWrapper>
        <TestInput
          name="test_input_name"
          disabled
          readOnly
        />
      </ContextWrapper>
    )

    const input = component.toJSON()
    expect(input).toMatchSnapshot()
  })

  it('changes state after onBlur event is triggered with no validation', async (done) => {
    await checkChange(done, 'handleOnBlur', {
      val: 'test_value',
      expected: {
        errors: { },
        warnings: { },
        isValid: true,
        eventName: 'onInputBlur',
        touched: true,
      },
    })
  })

  it('changes state after onBlur event is triggered with warn', async (done) => {
    await checkChange(done, 'handleOnBlur', {
      val: 'test_value',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      warn: [
        value => (value.length > 3 ? 'Warning message' : undefined),
      ],
      expected: {
        errors: { },
        warnings: { 0: 'Warning message' },
        isValid: true,
        eventName: 'onInputBlur',
        touched: true,
      },
    })
  })

  it('changes state after onBlur event is triggered with warn and error', async (done) => {
    await checkChange(done, 'handleOnBlur', {
      val: 'invalid_value',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      warn: [
        value => (value.length > 3 ? 'Warning message' : undefined),
      ],
      expected: {
        errors: { 0: 'Error message' },
        warnings: { 0: 'Warning message' },
        isValid: false,
        eventName: 'onInputBlur',
        touched: true,
      },
    })
  })

  it('changes state after onBlur event is triggered without warn or error', async (done) => {
    await checkChange(done, 'handleOnBlur', {
      val: 'ttt',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      warn: [
        value => (value.length > 3 ? 'Warning message' : undefined),
      ],
      expected: {
        errors: { },
        warnings: { },
        isValid: true,
        eventName: 'onInputBlur',
        touched: true,
      },
    })
  })

  it('shows no error if the input is disabled', async (done) => {
    await checkChange(done, 'handleOnBlur', {
      val: 'invalid_value',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      warn: [
        value => (value.length > 3 ? 'Warning message' : undefined),
      ],
      disabled: true,
      expected: {
        errors: { },
        warnings: { },
        isValid: true,
        eventName: 'onInputBlur',
        touched: true,
      },
    })
  })

  it('shows no error if the input is readOnly', async (done) => {
    await checkChange(done, 'handleOnBlur', {
      val: 'invalid_value',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      warn: [
        value => (value.length > 3 ? 'Warning message' : undefined),
      ],
      readOnly: true,
      expected: {
        errors: { },
        warnings: { },
        isValid: true,
        eventName: 'onInputBlur',
        touched: true,
      },
    })
  })

  it('will not async validate if there is sync error onBlur', async (done) => {
    await checkChange(done, 'handleOnBlur', {
      val: 'invalid_value',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      asyncValidation: value => fetchSimulate(1000).then(() => {
        if (value === 'invalid_value') {
          throw new Error('This name is forbidden')
        }
      }),
      expected: {
        errors: { 0: 'Error message' },
        warnings: { },
        isValid: false,
        eventName: 'onInputBlur',
        touched: true,
      },
    })
  })

  it('will show async error onBlur', async (done) => {
    await checkChange(done, 'handleOnBlur', {
      val: 'invalid_value_async',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      warn: [
        value => (value.length > 3 ? 'Warning message' : undefined),
      ],
      asyncValidation: value => fetchSimulate(10).then(() => {
        if (value === 'invalid_value_async') {
          throw new Error('This name is forbidden')
        }
      }),
      expected: {
        errors: { async: 'This name is forbidden' },
        warnings: { 0: 'Warning message' },
        isValid: false,
        eventName: 'onInputBlur',
        touched: true,
        asyncMessage: 'This name is forbidden',
      },
    })
  })

  it('changes state after onChange event is triggered with no validation', async (done) => {
    await checkChange(done, 'handleOnChange', {
      val: 'test_value',
      expected: {
        errors: { },
        warnings: { },
        isValid: true,
        eventName: 'onInputChange',
        touched: false,
      },
    })
  })

  it('changes state after onChange event is triggered with warn', async (done) => {
    await checkChange(done, 'handleOnChange', {
      val: 'test_value',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      warn: [
        value => (value.length > 3 ? 'Warning message' : undefined),
      ],
      expected: {
        errors: { },
        warnings: { 0: 'Warning message' },
        isValid: true,
        eventName: 'onInputChange',
        touched: false,
      },
    })
  })

  it('changes state after onChange event is triggered with warn and error', async (done) => {
    await checkChange(done, 'handleOnChange', {
      val: 'invalid_value',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      warn: [
        value => (value.length > 3 ? 'Warning message' : undefined),
      ],
      expected: {
        errors: { 0: 'Error message' },
        warnings: { 0: 'Warning message' },
        isValid: false,
        eventName: 'onInputChange',
        touched: false,
      },
    })
  })

  it('changes state after onChange event is triggered without warn or error', async (done) => {
    await checkChange(done, 'handleOnChange', {
      val: 'ttt',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      warn: [
        value => (value.length > 3 ? 'Warning message' : undefined),
      ],
      expected: {
        errors: { },
        warnings: { },
        isValid: true,
        eventName: 'onInputChange',
        touched: false,
      },
    })
  })

  it('shows no error if the input is disabled at onChange event', async (done) => {
    await checkChange(done, 'handleOnChange', {
      val: 'invalid_value',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      warn: [
        value => (value.length > 3 ? 'Warning message' : undefined),
      ],
      disabled: true,
      expected: {
        errors: { },
        warnings: { },
        isValid: true,
        eventName: 'onInputChange',
        touched: false,
      },
    })
  })

  it('shows no error if the input is readOnly at onChange event', async (done) => {
    await checkChange(done, 'handleOnChange', {
      val: 'invalid_value',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      warn: [
        value => (value.length > 3 ? 'Warning message' : undefined),
      ],
      readOnly: true,
      expected: {
        errors: { },
        warnings: { },
        isValid: true,
        eventName: 'onInputChange',
        touched: false,
      },
    })
  })

  it('will not async validate if there is sync error onChange', async (done) => {
    await checkChange(done, 'handleOnChange', {
      val: 'invalid_value',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      asyncValidation: value => fetchSimulate(1000).then(() => {
        if (value === 'invalid_value') {
          throw new Error('This name is forbidden')
        }
      }),
      expected: {
        errors: { 0: 'Error message' },
        warnings: { },
        isValid: false,
        eventName: 'onInputChange',
        touched: false,
      },
    })
  })

  it('will not async validate onChange', async (done) => {
    await checkChange(done, 'handleOnChange', {
      val: 'invalid_value_async',
      validate: [
        value => (value === 'invalid_value' ? 'Error message' : undefined),
      ],
      warn: [
        value => (value.length > 3 ? 'Warning message' : undefined),
      ],
      asyncValidation: value => fetchSimulate(10).then(() => {
        if (value === 'invalid_value_async') {
          throw new Error('This name is forbidden')
        }
      }),
      expected: {
        errors: { },
        warnings: { 0: 'Warning message' },
        isValid: true,
        eventName: 'onInputChange',
        touched: false,
      },
    })
  })

  it('removes async error onChange', async (done) => {
    const mockListenTo = jest.fn()
    const mockDispatchEvent = jest.fn()

    const component = shallow(
      (
        <TestInput
          name="test_input_name"
        />
      ),
      {
        context: {
          listenTo: mockListenTo,
          dispatchEvent: mockDispatchEvent,
        },
      }
    )

    component.setState({ errors: { async: 'some async error' } })
    expect(component.state('errors')).toEqual({ async: 'some async error' })

    const instance = component.instance()
    await instance.handleOnChange('some_value')

    expect(component.state('errors')).toEqual({ })

    done()
  })

  it('returns validation state onFormSubmit event', async (done) => {
    const mockDispatchEvent = jest.fn()
    const listeners = {}
    const listenTo = (eventName, method) => {
      listeners[eventName] = method
    }

    const component = shallow(
      (
        <TestInput
          name="test_input_name"
          validate={[
            value => (value === 'wrong_value' ? 'error' : undefined),
          ]}
        />
      ),
      {
        context: {
          listenTo,
          dispatchEvent: mockDispatchEvent,
        },
      }
    )

    component.setState({ value: 'wrong_value' })
    let result = await listeners.onFormSubmit()
    expect(result).toEqual({
      errors: { 0: 'error' },
      inputName: 'test_input_name',
      value: 'wrong_value',
      warnings: {},
    })

    component.setState({ value: 'test_value' })
    result = await listeners.onFormSubmit()
    expect(result).toEqual({
      errors: {},
      inputName: 'test_input_name',
      value: 'test_value',
      warnings: {},
    })

    done()
  })
})
