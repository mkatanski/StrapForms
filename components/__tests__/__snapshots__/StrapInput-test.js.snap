// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StrapInput can be rendered without problems as disabled and readonly 1`] = `
<div>
  <div>
    <input
      className={undefined}
      disabled={true}
      id="test_input_name"
      name="test_input_name"
      onBlur={[Function]}
      onChange={[Function]}
      readOnly={true}
      value=""
    />
    <span>
      {"input":{"disabled":true,"id":"test_input_name","name":"test_input_name","readOnly":true,"value":""},"hasErrors":false,"hasWarnings":false,"errors":{},"warnings":{},"touched":false,"isValidating":false,"children":null}
    </span>
  </div>
</div>
`;

exports[`StrapInput can be rendered without problems with default values 1`] = `
<div>
  <div>
    <input
      className={undefined}
      disabled={false}
      id="test_input_name"
      name="test_input_name"
      onBlur={[Function]}
      onChange={[Function]}
      readOnly={false}
      value=""
    />
    <span>
      {"input":{"disabled":false,"id":"test_input_name","name":"test_input_name","readOnly":false,"value":""},"hasErrors":false,"hasWarnings":false,"errors":{},"warnings":{},"touched":false,"isValidating":false,"children":null}
    </span>
  </div>
</div>
`;

exports[`StrapInput can be rendered without problems with initial value 1`] = `
<div>
  <div>
    <input
      className="customProperty"
      disabled={false}
      id="test_input_name"
      name="test_input_name"
      onBlur={[Function]}
      onChange={[Function]}
      readOnly={false}
      value="init value"
    />
    <span>
      {"className":"customProperty","input":{"disabled":false,"id":"test_input_name","name":"test_input_name","readOnly":false,"value":"init value"},"hasErrors":false,"hasWarnings":false,"errors":{},"warnings":{},"touched":false,"isValidating":false,"children":null}
    </span>
  </div>
</div>
`;
